File: packages/turbo-vlc/ios/Heyo.m
//
//  Heyo.m
//  react-native-turbo-vlc
//
//  Created by timothee on 24/09/2024.
//

#import <Foundation/Foundation.h>
#import <VLCKit/VLCKit.h>
#import "Heyo.h"

@implementation Heyo

- (instancetype)initWithFrame:(CGRect)frame
                         view:(UIView *)view
                      onError:(EventCallback)onError
                   onProgress:(EventCallback)onProgress
                   onVideoInfo:(EventCallback)onVideoInfo
                     onBuffer:(EventCallback)onBuffer {
    self = [super initWithFrame:frame];
    if (self) {
        self.vlc = [[VLCLibrary alloc] init];
        self.mediaPlayer = [[VLCMediaPlayer alloc] initWithLibrary:self.vlc];
        self.mediaPlayer.scaleFactor = 0;
        
        self.onError = onError;
        self.onProgress = onProgress;
        self.onVideoInfo = onVideoInfo;
        self.onBuffer = onBuffer;
        
        self.released = NO;
        self.mediaPlayer.delegate = self;
        self.mediaPlayer.drawable = self;
    }
    return self;
}

- (void)dealloc {
    NSLog(@"Objective-C dealloc");
}

- (void)releasePlayer {
    self.released = YES;
    [self.mediaPlayer stop];
}

- (void)prepare {
    self.released = NO;
}

- (BOOL)isReleased {
    return self.released;
}

#pragma mark - VLCMediaPlayerDelegate

- (void)mediaPlayerStateChanged:(VLCMediaPlayerState)aNotification {
    NSLog(@"mediaPlayerStateChanged");
}

- (void)mediaPlayerLengthChanged:(int64_t)aNotification {
    NSLog(@"mediaPlayerLengthChanged");
}

- (void)mediaPlayerTrackSelected:(NSString *)selectedId unselectedId:(NSString *)unselectedId {
    NSLog(@"mediaPlayerTrackSelected");
}

- (void)mediaPlayerTrackAdded:(NSString *)trackId {
    NSLog(@"mediaPlayerTrackAdded");
}

- (void)mediaPlayerTimeChanged:(NSNotification *)aNotification {
    if (self.released) return;
    
    NSDictionary *progress = @{
        @"progress": @((self.mediaPlayer.time.value ? self.mediaPlayer.time.value.doubleValue : 0)),
        @"duration": @((self.mediaPlayer.media.length.value ? self.mediaPlayer.media.length.value.doubleValue : 0) / 1000)
    };
    
    if (self.onProgress) {
        self.onProgress(progress);
    }
}

- (NSArray *)getTracks:(NSArray<VLCMediaPlayerTrack *> *)tracks {
    NSMutableArray *result = [NSMutableArray array];
    for (VLCMediaPlayerTrack *track in tracks) {
        [result addObject:@{
            @"id": track.trackId,
            @"name": [NSString stringWithFormat:@"%@ %@", track.trackName, track.trackDescription ?: @""]
        }];
    }
    return result;
}

#pragma mark - VLCMediaDelegate

- (void)mediaMetaDataDidChange:(VLCMedia *)aMedia {
    if (aMedia.length.intValue == 0 || self.released) return;

    NSDictionary *videoInfo = @{
        @"duration": @((self.mediaPlayer.media.length.value.doubleValue) / 1000),
        @"audioTracks": [self getTracks:self.mediaPlayer.audioTracks],
        @"videoTracks": [self getTracks:self.mediaPlayer.videoTracks],
        @"textTracks": [self getTracks:self.mediaPlayer.textTracks],
        @"currentVideoTrackId": [self getSelectedTrackId:self.mediaPlayer.videoTracks],
        @"currentAudioTrackId": [self getSelectedTrackId:self.mediaPlayer.audioTracks],
        @"currentTextTrackId": [self getSelectedTrackId:self.mediaPlayer.textTracks]
    };
    
    if (self.onVideoInfo) {
        self.onVideoInfo(videoInfo);
    }
}

- (NSString *)getSelectedTrackId:(NSArray<VLCMediaPlayerTrack *> *)tracks {
    for (VLCMediaPlayerTrack *track in tracks) {
        if (track.isSelected) {
            return track.trackId;
        }
    }
    return @"";
}

- (void)mediaDidFinishParsing:(VLCMedia *)aMedia {
    NSLog(@"mediaDidFinishParsing");
}

#pragma mark - Setters

- (void)setVolume:(int32_t)volume {
    self.mediaPlayer.audio.volume = volume;
}

- (void)setUri:(NSString *)uri {
    NSURL *parsedUri = [NSURL URLWithString:uri];
    if (!parsedUri) return;

    VLCMedia *media = [[VLCMedia alloc] initWithURL:parsedUri];
    if (!media) return;

    if (self.media) {
        self.media.delegate = nil;
    }

    self.media = media;
    self.media.delegate = self;
    self.mediaPlayer.media = media;
    
    NSLog(@"PLAYING");
    [self.mediaPlayer play];
}

- (void)setPlay:(BOOL)play {
    NSLog(@"Updating play %d", play);
    if (!self.mediaPlayer.media) return;

    if (play) {
        [self.mediaPlayer play];
    } else {
        [self.mediaPlayer pause];
    }
}

- (void)setSeek:(double)seek {
    self.mediaPlayer.time = [VLCTime timeWithInt:(int32_t)seek];
}

- (void)setAudioTrack:(NSString *)trackId {
    for (VLCMediaPlayerTrack *track in self.mediaPlayer.audioTracks) {
        track.selected = [track.trackId isEqualToString:trackId];
    }
    
    if ([trackId isEqualToString:@"default"]) {
        VLCMediaPlayerTrack *firstTrack = self.mediaPlayer.audioTracks.firstObject;
        if (firstTrack) {
            firstTrack.selected = YES;
        }
    }
}

- (void)setTextTrack:(NSString *)trackId {
    for (VLCMediaPlayerTrack *track in self.mediaPlayer.textTracks) {
        track.selected = [track.trackId isEqualToString:trackId];
    }
    
    if ([trackId isEqualToString:@"default"]) {
        VLCMediaPlayerTrack *firstTrack = self.mediaPlayer.textTracks.firstObject;
        if (firstTrack) {
            firstTrack.selected = YES;
        }
    }
}

@end

File: packages/turbo-vlc/ios/Heyo.swift
import Foundation
import UIKit

public typealias EventCallback = (_ event: NSDictionary) -> Void

@objcMembers
public class Heyo: UIView, VLCMediaPlayerDelegate, VLCMediaDelegate {
  private var vlc: VLCLibrary
  private var mediaPlayer: VLCMediaPlayer
  private var media: VLCMedia?
  
  private var onError: EventCallback
  private var onProgress: EventCallback
  private var onVideoInfo: EventCallback
  private var onBuffer: EventCallback
  
  private var released = false
  
  @objc
  public init(frame: CGRect, view: UIView, onError: @escaping EventCallback, onProgress: @escaping EventCallback, onVideoInfo: @escaping EventCallback, onBuffer: @escaping EventCallback) {
        
    self.vlc = VLCLibrary()
    self.mediaPlayer = VLCMediaPlayer(library: self.vlc)
    
    self.mediaPlayer.scaleFactor = 0
    
    self.onError = onError
    self.onProgress = onProgress
    self.onVideoInfo = onVideoInfo
    self.onBuffer = onBuffer
    
    super.init(frame: frame)
    
    self.mediaPlayer.delegate = self
    self.mediaPlayer.drawable = self
  }
  
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  deinit {
    print("SWIFT deinit")
  }
  
  @objc
  public func releasePlayer() {
    self.released = true
    self.mediaPlayer.stop()
  }
  
  @objc func prepare() {
    self.released = false
  }
  
  @objc func isReleased() -> Bool {
    return self.released
  }
  
  public func mediaPlayerStateChanged(_ newState: VLCMediaPlayerState) {
    print("mediaPlayerStateChanged")
  }
  
  public func mediaPlayerLengthChanged(_ length: Int64) {
    print("mediaPlayerLengthChanged")
  }
  
  public func mediaPlayerTrackSelected(selectedId unselectedId: String, unselectedId unselectedId2: String) {
    print("mediaPlayerTrackSelected")
  }
  
  public func mediaPlayerTrackAdded(_ trackId: String) {
    print("mediaPlayerTrackAdded")
  }
  
  public func mediaPlayerTimeChanged(_ aNotification: Notification) {
    if self.released {
      return
    }
    self.onProgress([
      "progress": (self.mediaPlayer.time.value?.doubleValue ?? 0),
      "duration": (self.mediaPlayer.media?.length.value?.doubleValue ?? 0) / 1000,
    ])
  }
  
  private func getTracks(tracks: [VLCMediaPlayer.Track]) -> NSMutableArray {
    let a: NSMutableArray = []
    tracks.forEach({ track in
      a.add([
        "id": track.trackId,
        "name": "\(track.trackName) \(track.trackDescription ?? "")",
      ])
    })
    return a
  }
  
  public func mediaMetaDataDidChange(_ aMedia: VLCMedia) {
    if aMedia.length.intValue == 0 {
      return
    }
    if self.released {
      return
    }
    self.onVideoInfo([
      "duration": (self.mediaPlayer.media?.length.value?.doubleValue ?? 0) / 1000,
      
      "audioTracks": self.getTracks(tracks: self.mediaPlayer.audioTracks),
      "videoTracks": self.getTracks(tracks: self.mediaPlayer.videoTracks),
      "textTracks": self.getTracks(tracks: self.mediaPlayer.textTracks),
      
      "currentVideoTrackId": self.mediaPlayer.videoTracks.first(where: { track in track.isSelected })?.trackId ?? "",
      "currentAudioTrackId": self.mediaPlayer.audioTracks.first(where: { track in track.isSelected })?.trackId ?? "",
      "currentTextTrackId": self.mediaPlayer.textTracks.first(where: { track in track.isSelected })?.trackId ?? "",
    ])
  }
  
  public func mediaDidFinishParsing(_ aMedia: VLCMedia) {
    print("mediaDidFinishParsing")
  }
    
  @objc
  public func setVolume(volume: Int32) {
    self.mediaPlayer.audio?.volume = volume
  }
  
  @objc
  public func setUri(uri: String) {
    guard let parsedUri = URL.init(string: uri) else { return }
    guard let media = VLCMedia(url: parsedUri) else { return }

    if let m = self.media {
      m.delegate = nil
    }
    
    self.media = media
    media.delegate = self
    self.mediaPlayer.media = self.media

    print("PLAYING")
    
    self.mediaPlayer.play()
  }
  
  @objc
  public func setPlay(play: Bool) {
    print("Updating play \(play)")
    guard (self.mediaPlayer.media != nil) else { return }
    print("Updating play after \(play)")
    if play {
      self.mediaPlayer.play()
    } else {
      self.mediaPlayer.pause()
    }
  }
  
  @objc
  public func setSeek(seek: Double) {
    self.mediaPlayer.time = VLCTime(int: Int32(seek))
  }
  
  @objc
  public func setAudioTrack(id: String) {
    self.mediaPlayer.videoTracks.forEach { track in
      track.isSelected = track.trackId == id
    }
    if id == "default" {
      if let first = self.mediaPlayer.videoTracks.first {
        first.isSelected = true
      }
    }
  }
  
  @objc
  public func setTextTrack(id: String) {
    self.mediaPlayer.textTracks.forEach { track in
      track.isSelected = track.trackId == id
    }
    if id == "default" {
      if let first = self.mediaPlayer.textTracks.first {
        first.isSelected = true
      }
    }
  }
}
